(3/8) [REFACTOR] P2PDataStore::add/remove/refresh

julianknutsen opened the pull request:
Improve clarity and add comments for future readers to help them reason about the behavior.

This also paves the way for deduplicating the remove() paths and helps identify the shared code paths
that can be merged in future refactoring work.

julianknutsen committed:
[REFACTOR] P2PDataStorage::addPersistableNetworkPayload()

Add comments and refactor the body in order to make it easier to
understand.

julianknutsen committed:
[BUGFIX] Don't try and remove() if addMailboxData() fails

Fix a bug where remove() was called in the addMailboxData()
failure path.

1. Sender's can't remove mailbox entries. Only
   the receiver can remove it so even if the previous add() failed and
   left partial state, the remove() can never succeed.

2. Even if the sender could remove, this path used remove() instead
   of removeMailboxData() so it wouldn't have succeed anyway.

This patch cleans up the failure path as well as adds a precondition
for the remove() function to ensure future callers don't use them for
ProtectedMailboxStorageEntrys.

julianknutsen committed:
[REFACTOR] P2PDataStorage::addProtectedStorageEntry()

Refactor addProtectedStorageEntry for more readability and add comments
to help future readers.

julianknutsen committed:
[REFACTOR] P2PDataStorage::refreshTTL()

Refactor for readability and add comments for future readers.

julianknutsen committed:
[REFACTOR] P2PDataStorage::remove()

Refactor for readability and add comments to help future readers.

julianknutsen committed:
[REFACTOR] P2PDataStorage::removeMailboxData()

Refactor for readability and add comments for future readers.

chimp1984 commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -316,33 +316,41 @@ public boolean addPersistableNetworkPayload(PersistableNetworkPayload payload,
+        // Payload hash size does not match expectation for that type of message.
+        if (!payload.verifyHashSize()) {
             log.warn("We got a hash exceeding our permitted size");
             return false;
```
I think we do not check for exceeding, just if hash size is as expected.

julianknutsen replied:
This comment has been updated.

chimp1984 commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -390,50 +390,64 @@ public boolean addProtectedStorageEntry(ProtectedStorageEntry protectedStorageEn
+                !checkSignature(protectedStorageEntry)) {
+
+            log.trace("addProtectedStorageEntry failed due to invalid entry");
+            return false;
```
Maybe better to split that to 2 checks and more concrete log messages.

julianknutsen replied:
The helper methods already have verbose logging and each case will print out a separate message. I updated this PR to only add new messages for the failures that didn't have them before. It is worth another pass to clean these up again, but I don't want to pollute this stack with those changes.

chimp1984 commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -390,50 +390,64 @@ public boolean addProtectedStorageEntry(ProtectedStorageEntry protectedStorageEn
-            result = result && checkIfStoredDataPubKeyMatchesNewDataPubKey(protectedStorageEntry.getOwnerPubKey(), hashOfPayload);
+
+        // In a hash collision between two well formed ProtectedStorageEntry, the first item wins and will not be overwritten
+        if (containsKey &&
```
> In a hash collision between two well formed ProtectedStorageEntry, the first item wins and will not be overwritten
       
I think that comment is misleading. The checkIfStoredDataPubKeyMatchesNewDataPubKey methods just checks if the pubKey is the same and if the hash is matching. Hash collusion is not an issue as we use cryptographic hashes as keys.

julianknutsen replied:
Do you have much context on when this condition is expected?

My understanding of the check is to verify that the NewEntry.ownerPubkKey == StoredEntry.ownerPubKey, but how could this fail unless there were identical SeqNr & Payload hashes from different owners? Maybe a misbehaving node sets the value wrong?

I may be missing something really straight forward, but I am happy to update the comment to be more appropriate for the check and not focus on hash collisions.

julianknutsen replied:
This has been renamed to matchesRelevantPubKey in later patches. The idea is that for ProtectedStorageEntrys we want the Entry.ownerPubKey to match the Payload.ownerPubKey.

For ProtectedMailboxStorageEntrys we want the Entry.receiversPubKey to match the Payload.ownerPubKey.

chimp1984 commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -390,50 +390,64 @@ public boolean addProtectedStorageEntry(ProtectedStorageEntry protectedStorageEn
+        if (containsKey &&
+                !checkIfStoredDataPubKeyMatchesNewDataPubKey(protectedStorageEntry.getOwnerPubKey(), hashOfPayload)) {
+
+            log.trace("addProtectedStorageEntry failed due to hash collision");
```
See above comment...

chimp1984 commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -461,39 +461,64 @@ public boolean refreshTTL(RefreshOfferMessage refreshTTLMessage,
+        // TODO: I think we can return false here. All callers use the get() public API which increments the sequence number
+        // leaving only the onMessage() handler which doesn't look at the return value. It makes more intuitive sense that operations that don't
+        // change state return false.
+        if (sequenceNumberMap.containsKey(hashOfPayload) && sequenceNumberMap.get(hashOfPayload).sequenceNr == sequenceNumber) {
```
In OfferBookService it is used and would cause a error message in that case (thought I think its an impossible case). Have not followed further if that error message would cause more troubles but I think it would require more checks of all client usage code to be sure we can change return value.  Conceptually I agree.

chimp1984 commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -390,50 +390,64 @@ public boolean addProtectedStorageEntry(ProtectedStorageEntry protectedStorageEn
-                log.trace("We got that version of the data already, so we don't store it.");
-            }
+        // If we have seen a more recent operation for this payload, we ignore the current one
+        // TODO: I think we can return false here. All callers use the get() public API which increments the sequence number
```
There is quite a bit of client code dealing with the return value. Would require more analysis if we can safely change it. Conceptually I agree.

julianknutsen replied:
I agree. I'll add my thoughts about this on (4/4) to explain more about the current state and my opinion on where it should go to make it a bit easier to reason about. I agree that it is possible to client code may expect the current behavior and I will dig into each path individually to understand what happens on failure.

julianknutsen committed:
[PR COMMENTS] Clean up logging messages

Removed duplicate log messages that are handled inside the various helper methods
and print more verbose state useful for debugging.

Updated potentially misleading comments around hashing collisions

julianknutsen commented:
Commits start at: 5512f34

This got stale and polluted with merge commits. Updated against master and all previous comments have been addressed.

freimair commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -380,50 +390,50 @@ public boolean addProtectedStorageEntry(ProtectedStorageEntry protectedStorageEn
-                sequenceNumberMap.put(hashOfPayload, new MapValue(protectedStorageEntry.getSequenceNumber(), System.currentTimeMillis()));
-                // We set the delay higher as we might receive a batch of items
-                sequenceNumberMapStorage.queueUpForSave(SequenceNumberMap.clone(sequenceNumberMap), 2000);
+        // If we have seen a more recent operation for this payload, we ignore the current one
```
```suggestion
        // If we have seen a more recent version of this payload, we ignore the current one
```

?

julianknutsen replied:
We don't necessarily see a different version of the payload as the contents are identical, but we do see different operations on that payload that increment the sequence number.

The producers of change (add/remove/refresh) always increment the SeqNr so it acts like a timestamp for which operations take precedence over one another. For example, a Remove operation with SeqNr 2 is superseded by an add with SeqNr 3.

In that way of thinking, it made sense to me that the SeqNrs are more tied to the operations that occur and less about the payload version.

Hope that helps.

freimair reviewed (approved):
utAck

ripcurlx reviewed (approved):
utACK - 👍 makes it way easier to read than before.
